### PR TITLE
BUGFIX: SD-460 Combo chart - Missing data labels

### PR DESCRIPTION
**Pull request by nadtan**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2250
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2002

# Related Jira tasks

### CODE REVIEW

**Comment by nadtan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3678/ : SUCCESS in 4m 37s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4589/ : SUCCESS in 34s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1319/ : FAILURE in 1h 00m 22s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4590/ : SUCCESS in 10s

**Comment by nadtan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3737/ : SUCCESS in 7m 56s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4688/ : SUCCESS in 10s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1363/ : FAILURE in 10m 29s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4689/ : SUCCESS in 8s

**Comment by nadtan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by nadtan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1388/ : FAILURE in 10m 43s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4741/ : SUCCESS in 13s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3765/ : SUCCESS in 11m 37s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4739/ : SUCCESS in 1m 04s

**Comment by nadtan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3793/ : SUCCESS in 5m 19s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4786/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1411/ : FAILURE in 11m 48s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4787/ : SUCCESS in 50s

**Comment by nadtan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1416/ : FAILURE in 13m 59s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4795/ : SUCCESS in 9s

**Comment by nadtan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3810/ : SUCCESS in 15m 29s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4830/ : SUCCESS in 11s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1438/ : ABORTED in 40m 16s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4831/ : SUCCESS in 11s

**Comment by nadtan** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2085/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1440/ : SUCCESS in 11m 38s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4834/ : SUCCESS in 18s

**Comment by nadtan** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-992.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1441/ : SUCCESS in 46m 08s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4836/ : SUCCESS in 10s

File: src/components/visualizations/chart/highcharts/test/dataLabelsHelpers.spec.ts [hunk 1]
```diff
@@ -343,4 +346,38 @@ describe("dataLabelsHelpers", () => {
...
+
+    describe("getLabelStyle", () => {
+        const CHART_TYPES = [
+            VisualizationTypes.COLUMN,
+            VisualizationTypes.BAR,
+            VisualizationTypes.LINE,
+            VisualizationTypes.COMBO2,
+        ];
+
+        it.each([null, "normal", "percent"])(
```
[the remark this answers is not in the document]
Comment (nadtan) [reply src/components/visualizations/chart/highcharts/test/dataLabelsHelpers.spec.ts]: updated

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3814/ : SUCCESS in 4m 34s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4840/ : SUCCESS in 11s

**Comment by naibinh** [comment]
LGTM

File: src/components/visualizations/chart/highcharts/dataLabelsHelpers.ts [hunk 2]
```diff
@@ -149,3 +156,41 @@ export function getShapeVisiblePart(shape: any, chart: any, wholeSize: number) {
...
+
+export function getLabelStyle(type: string, stacking: string) {
+    if (isAreaChart(type)) {
+        return BLACK_LABEL;
+    }
+    return stacking || isOneOfTypes(type, whiteDataLabelTypes) ? WHITE_LABEL : BLACK_LABEL;
+}
+
+export function formatAsPercent(unit: number = 100): string {
+    const val = parseFloat((this.value * unit).toPrecision(14));
```
Comment (martinmosko) [inline src/components/visualizations/chart/highcharts/dataLabelsHelpers.ts]: I know you just moved the existing code but please `this.` here!!! not inside class? 

please create a ticket for this to fix this function without using "this"

also this place: https://github.com/gooddata/gooddata-react-components/blob/481391606d649919c65af06383cdd7d623e192d3/src/components/visualizations/chart/highcharts/plugins/dualAxesLabelFormatter.ts#L122 
without using `this` please

File: src/constants/label.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// (C) 2019 GoodData Corporation
+import { VisualizationTypes } from "./visualizationTypes";
+
+export const WHITE_LABEL = {
```
Comment (martinmosko) [inline src/constants/label.ts]: missing type ? :)

File: src/components/visualizations/chart/highcharts/dataLabelsHelpers.ts (hunk 2 above)
Comment (nadtan) [reply src/components/visualizations/chart/highcharts/dataLabelsHelpers.ts]: Reported in https://jira.intgdc.com/browse/SD-484

File: src/constants/label.ts (hunk 3 above)
Comment (nadtan) [reply src/constants/label.ts]: added ;)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3844/ : SUCCESS in 7m 30s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4870/ : SUCCESS in 11s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3845/ : SUCCESS in 5m 15s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4872/ : SUCCESS in 12s

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2102/ : SUCCESS in 10m 15s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3846/ : SUCCESS in 13m 32s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1015/ : SUCCESS in 27s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1162/ : FAILURE in 34m 17s

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2103/ : SUCCESS in 10m 19s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3848/ : SUCCESS in 12m 36s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1016/ : SUCCESS in 22s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1163/ : FAILURE in 21m 20s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2105/ : SUCCESS in 12m 25s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3851/ : SUCCESS in 7m 28s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1018/ : SUCCESS in 16s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1165/ : SUCCESS in 26m 45s

**Pull request merged**